src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs

```diff
@@ -25,20 +29,81 @@ public class WalletRPCController : FeatureController
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
         }
 
-        [ActionName("sendtoaddress")]
-        [ActionDescription("Sends money to a bitcoin address.")]
-        public uint256 SendToAddress(BitcoinAddress bitcoinAddress, Money amount)
+        [ActionName("walletpassphrase")]
+        [ActionDescription("Stores the wallet decryption key in memory for the indicated number of seconds. Issuing the walletpassphrase command while the wallet is already unlocked will set a new unlock time that overrides the old one.")]
+        [NoTrace]
+        public bool UnlockWallet(string passphrase, int timeout)
+        {
+            Guard.NotEmpty(passphrase, nameof(passphrase));
+
+            // As per RPC method definition this should be the max allowable expiry duration.
+            const int maxDurationInSeconds = 1073741824;
+
+            WalletAccountReference account = this.GetAccount();
+
+            // Length of expiry of the unlocking, restricted to max duration.
+            TimeSpan duration = new TimeSpan(0, 0, Math.Min(timeout, maxDurationInSeconds));
+
+            try
+            {
+                this.walletTransactionHandler.CacheSecret(account, passphrase, duration);
+            }
+            catch (SecurityException exception)
+            {
+                throw new RPCServerException(RPCErrorCode.RPC_INVALID_REQUEST, exception.Message);
+            }
+            return true; // NOTE: Have to return a value or else RPC middleware doesn't serialize properly.
+        }
+
+        [ActionName("walletlock")]
+        [ActionDescription("Removes the wallet encryption key from memory, locking the wallet. After calling this method, you will need to call walletpassphrase again before being able to call any methods which require the wallet to be unlocked.")]
+        public bool LockWallet()
         {
             WalletAccountReference account = this.GetAccount();
-            return uint256.Zero;
+            this.walletTransactionHandler.ClearCachedSecret(account);
+            return true; // NOTE: Have to return a value or else RPC middleware doesn't serialize properly.
+        }
+
+        [ActionName("sendtoaddress")]
+        [ActionDescription("Sends money to a bitcoin address. Requires wallet to be unlocked using walletpassphrase.")]
+        public async Task<uint256> SendToAddressAsync(BitcoinAddress address, decimal amount, string commentTx, string commentDest)
+        {
+            WalletAccountReference account = this.GetAccount(); 
+            TransactionBuildContext context = new TransactionBuildContext(this.fullNode.Network)
+            {
+                AccountReference = this.GetAccount(),
+                Recipients = new [] {new Recipient { Amount = Money.Coins(amount), ScriptPubKey = address.ScriptPubKey } }.ToList(),
+                WalletPassword = "_" // Want private key to pull from cache, so pass in non empty string so tx will be signed.
```

Comment:
This feels a bit ugly

Reply:
yeah it is ugly just use an empty string

Reply:
Well, the empty-ness is used inside `BuildTransaction` to check whether signing should occur. But there should be a less hacky solution than this. Maybe a future PR could require an explicit bool in `TransactionBuildContext` for signing.

Reply:
Yes zeptin is correct, I would have to either add a parameter to the context, add a different method or add a parameter to the method. Not sure which is best so will wait to see what Jeremy thinks about this code.

Reply:
Ok I see the argument, the correct approach is a flag that says to look in cache or use password.
But for now I guess this is ok.

Reply:
Perhaps this code should go in the else if key is not in cache.
```
            if (string.IsNullOrEmpty(context.WalletPassword))
                return;
```

Reply:
This is a hard one. I'm happy for this to go in like this, but @mikedennis can you please follow with a PR that adds a `bool Sign` to the TransactionBuildContext? True by default.
If sign = true and there is no password then we rely on the cache, if sign is false and there is password we ignore it etc.

Reply:
Yes for sure. Sounds like a good solution.